pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -472,91 +472,124 @@ func ResolvePipelineRunTask(
+	if len(pipelineTask.Conditions) > 0 {
+		rcc, err := resolveConditionChecks(&pipelineTask, pipelineRun.Status.TaskRuns, pipelineRun.Status.ChildReferences, rprt.TaskRunName, getTaskRun, getCondition, providedResources)
 		if err != nil {
-			return nil, fmt.Errorf("couldn't match referenced resources with declared resources: %w", err)
+			return err
 		}
+		rprt.ResolvedConditionChecks = rcc
+	}
+	return nil
+}
 
-		rprt.ResolvedTaskResources = rtr
+func resolveTaskResources(
+	ctx context.Context,
+	getTask resources.GetTask,
+	pipelineTask v1beta1.PipelineTask,
+	providedResources map[string]*resourcev1alpha1.PipelineResource,
+	taskRun *v1beta1.TaskRun,
+) (*resources.ResolvedTaskResources, error) {
 
-		// Get all conditions that this pipelineTask will be using, if any
-		if len(task.Conditions) > 0 {
-			rcc, err := resolveConditionChecks(&task, pipelineRun.Status.TaskRuns, pipelineRun.Status.ChildReferences, rprt.TaskRunName, getTaskRun, getCondition, providedResources)
-			if err != nil {
-				return nil, err
-			}
-			rprt.ResolvedConditionChecks = rcc
-		}
+	spec, taskName, kind, err := resolveTask(ctx, taskRun, getTask, pipelineTask)
+	if err != nil {
+		return nil, err
 	}
-	return &rprt, nil
+
+	spec.SetDefaults(ctx)
+	rtr, err := resolvePipelineTaskResources(pipelineTask, &spec, taskName, kind, providedResources)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't match referenced resources with declared resources: %w", err)
+	}
+
+	return rtr, nil
 }
 
-func resolveTask(ctx context.Context, taskRun *v1beta1.TaskRun, getTask resources.GetTask, task v1beta1.PipelineTask) (v1beta1.TaskSpec, string, v1beta1.TaskKind, error) {
+func resolveTask(
+	ctx context.Context,
+	taskRun *v1beta1.TaskRun,
+	getTask resources.GetTask,
+	pipelineTask v1beta1.PipelineTask,
+) (v1beta1.TaskSpec, string, v1beta1.TaskKind, error) {
 	var (
 		t        v1beta1.TaskObject
 		err      error
 		spec     v1beta1.TaskSpec
 		taskName string
 		kind     v1beta1.TaskKind
 	)
-	if task.TaskRef != nil {
+
+	if pipelineTask.TaskRef != nil {
 		// If the TaskRun has already a stored TaskSpec in its status, use it as source of truth
 		if taskRun != nil && taskRun.Status.TaskSpec != nil {
 			spec = *taskRun.Status.TaskSpec
-			taskName = task.TaskRef.Name
+			taskName = pipelineTask.TaskRef.Name
 		} else {
-			t, err = getTask(ctx, task.TaskRef.Name)
+			t, err = getTask(ctx, pipelineTask.TaskRef.Name)
 			switch {
 			case errors.Is(err, remote.ErrorRequestInProgress):
 				return v1beta1.TaskSpec{}, "", "", err
 			case err != nil:
 				return v1beta1.TaskSpec{}, "", "", &TaskNotFoundError{
-					Name: task.TaskRef.Name,
+					Name: pipelineTask.TaskRef.Name,
 					Msg:  err.Error(),
 				}
 			default:
 				spec = t.TaskSpec()
 				taskName = t.TaskMetadata().Name
 			}
 		}
-		kind = task.TaskRef.Kind
-	} else {
-		spec = task.TaskSpec.TaskSpec
+		kind = pipelineTask.TaskRef.Kind
+	} else if pipelineTask.TaskSpec != nil {
```

Comment:
This extra check will be much clear with a unit test. From what I recall, we have `webhook` validation to at least have either `taskRef` or `taskSpec` initialized. So, if `taskRef` is not initialized, its safe to rely on `taskSpec` without checking it being not-nil. At this point, we can avoid introducing such condition check specially with lack of a unit test.

Reply:
good point, removed that extra check 👍🏾